src/netrc.c

```diff
@@ -37,78 +37,127 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <stdbool.h>
-#include "axel.h"
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
 #include "netrc.h"
 
-enum lookup_state {
-	NOTHING,
-	DEFAULT,
-	MACHINE,
-	HOST_ENTRY,
-	LOGIN,
-	PASSWORD
-};
-
-void
-netrc_parse(conn_t *conn)
+#define REL_NETRC	"/.netrc"
+#define IS_DELIM(x)	(*x == ' ' || *x == '\t' || *x == '\n')
+
+static int fd;
+static size_t sz;
+static char *p;
+static char *s_addr;
+static char *e_addr;
+
+static size_t file_size(void)
+{
+	struct stat st;
+	fstat(fd, &st);
+	return st.st_size;
+}
+
+static int netrc_open(const char * const filename)
 {
-	FILE *fp;
-	bool found = false;
+	int ret;
+	char *path;
+	char *netrc;
 	char *home;
-	char *tok, *ntok;
-	char line[MAX_STRING];
-	char const *delims = " \t\n";
-	enum lookup_state state = NOTHING;
-
-	if (conn->conf->netrc_filename[0] == '\0') {
-		home = getenv("HOME");
-		snprintf(conn->conf->netrc_filename, MAX_STRING, "%s/.netrc", home);
+
+	ret = 0;
+	if (filename && filename[0] != '\0') {
+		path = strdup(filename);
+	} else {
+		if ((netrc = getenv("NETRC"))) {
+			path = strdup(netrc);
```

Comment:
Yes, that could be used, but we would have another special case to deal with, because the pointers returned by `getenv` wouldn't (and couldn't) be `free`d, but we'd have yet to `free` the allocated memory where we build the full path for the `.netrc` file.

I used strdup in this case to avoid dealing with those special cases.

Reply:
Ok, understood. I already made the changes to avoid `strdup`.